### DISCUSSION TITLE
After the group was removed, which API should be used as a replacement?

### DISCUSSION THREAD

**Discussion by phil-skillwon** [actor-framework/actor-framework#1965] [Q&A] [answered]
Starting from version caf-1.0.0, the group and related APIs have been removed. How can I implement group-based broadcasting now? For example, I need to send a command to collect data to all sensors within a group. How can I achieve this, and is there any simple code I can refer to?

**Comment by Neverlord** [comment]
What to use instead depends a bit on your use case and what you are sending. If flow-based processing makes sense, a `caf::stream` is a good replacement. Alternatively, actors can register explicitly at another actor for receiving certain messages. If you need n:m communication (n producers, m consumers), a broker actor maintaining a list of subscribers would probably be the way to go. If you need access to that broker from multiple different points in the application, storing it in the registry might be a good idea.

**Reply by phil-skillwon** [reply]
@Neverlord 
Thank you for your suggestions. I am inclined to choose this approach: create a subscription broker actor, and have the broker actor maintain a list of subscribers internally. The list would look something like this:
```C++
class MyGroup final: public caf::event_based_actor
{
    // ...
private:
    std::list<caf::actor> m_subscriberList;
    // ...
}
```
My understanding of the workflow for this approach is roughly as follows:

1. A subscriber sends a ctrl message to the subscription broker actor, sending its actor handle to the broker. Upon receiving the message, the broker actor calls monitor to track the subscriber's lifecycle. If the subscriber exits or unsubscribes, the broker actor will remove the subscriber from its internal list.
2. When the subscription broker actor receives a regular message, it traverses the internal subscriber list and broadcasts the received data to all subscribers.
3. When the subscription broker actor exits, it cancels monitoring of all subscribers and clears the internal subscriber list.
    
I'm not sure if the above description aligns with what you were trying to express earlier.
   

By the way, since the year before last, in addition to using CAF for some web and personal game projects, I have also used CAF for many embedded software projects, and I made extensive use of group. I find it to be a very useful and commonly used data structure. This shows that CAF has good performance in software projects in other industries as well. The last version of CAF I used was 0.19.1, and I only recently started using CAF 1.0.2. I would like to ask: What was the reasoning behind removing group and related APIs?

**Reply by Neverlord** [reply, marked as answer]
We've deprecated groups with 0.19 (from the changelog):

> The group-based pub/sub mechanism never fit nicely into the typed messaging API and the fact that group messages use the regular mailbox makes it hard to separate regular communication from multi-cast messages. Hence, we decided to drop the group API and instead focus on the new flows and streams that can replace group-communication in many use cases.

A bit more background:

Groups were meant as loosely coupled pub/sub mechanism and the original design dates way back before the addition of typed actors. As a result, they interacted very poorly with typed actors (cf. https://github.com/actor-framework/actor-framework/issues/441 and https://github.com/actor-framework/actor-framework/issues/225).

The biggest design flaw of groups was that they use the regular actor mailbox for publishing items, which means groups had to be implemented at the same level as actors in the class hierarchy since `send` had to work on both. But you couldn't `request` to a group (https://github.com/actor-framework/actor-framework/issues/632), even if the group was local. By using regular actor messaging, there's also no builtin back-pressure for groups. A single slow subscriber in a group with a high message volume will blow up your application (by piling up messages in the mailbox until exhausting all memory).

A `stream` (or `typed_stream`) can replace groups for use cases where you only have a single producer (or external source like events from a WebSocket). They come with back-pressure, have an API that separates regular actors messages from stream items, have proper error handling, and have a statically typed version.

The only use case we currently lack a good API for is multi-producer, multi-consumer "groups". As long as your group is low volume, the broker actor you've outlined will do fine (except please don't subtype `event_based_actor` and use stateful actors instead). If you have publishers that can produce lots of items, flows/streams is the safer option again. The implementation could be a broker actor again, but this time it would not receive messages that it forwards, but streams that it merges together into one big stream that all subscribers listen to. Publishers slow down automatically due to the builtin back-pressure or you can use the `on_backpressure_buffer` operator to cut off slow subscribers instead, whichever makes more sense for your application. Either way, the end result is more robust and safer than something based on the old groups API.

**Reply by phil-skillwon** [reply]
@Neverlord 
Thank you, I have a few more questions:

1. Regarding the SPMC scenario previously described, I read examples/flow/spsc-buffer-resource.cpp, and then I tried to implement multicast functionality using caf::spsc_buffer. However, only one subscriber can receive the data, while the other subscriber cannot. I might need to use spmc_buffer, but CAF does not have this data structure. Are there any similar examples I can refer to?

2. About the examples in the examples directory, are there more reference codes available? Although there are many examples in the directory, it seems that not all functionalities of some CAF components are demonstrated. This is especially true for the usage instructions related to flow.

3. You previously mentioned: "If you have publishers that can produce lots of items, flows/streams is the safer option again. The implementation could be a broker actor again, but this time it would not receive messages that it forwards, but streams that it merges together into one big stream that all subscribers listen to." I couldn't find this in the examples directory. Could you provide some simple code to demonstrate how to implement this functionality?

4. I read this document: https://actor-framework.readthedocs.io/en/stable/core/DataFlows.html. I have some confusion regarding the relationship between observable, observer, and subscription. My understanding is as follows: the observer corresponds to the consumer, the observable corresponds to the producer, and the subscription is initiated by the observer calling the subscribe instance method. Is my understanding correct? But from examples/flow/spsc-buffer-resource.cpp, it doesn't seem to be what I understood.
```C++
// --(rst-source-begin)--
// Simple source for generating a stream of integers from 1 to n.
void source(caf::event_based_actor* self,
            caf::async::producer_resource<int> out, size_t n) {
  self
    // Get an observable factory.
    ->make_observable()
    // Produce an integer sequence starting at 1, i.e., 1, 2, 3, ...
    .iota(1)
    // Only take the requested number of items from the infinite sequence.
    .take(n)
    // Subscribe the resource to the sequence, thereby starting the stream.
    .subscribe(out);
}
// --(rst-source-end)--

// --(rst-sink-begin)--
// Simple sink for consuming a stream of integers, printing it to stdout.
void sink(caf::event_based_actor* self, caf::async::consumer_resource<int> in) {
  self
    // Get an observable factory.
    ->make_observable()
    // Lift the input to an observable flow.
    .from_resource(std::move(in))
    // Print each integer.
    .for_each([self](int x) { self->println("{}", x); });
}
// --(rst-sink-end)--
```

**Reply by Neverlord** [reply]
1. I believe the API you are looking for is the `publisher`.
2. See 4.
3. It's just calling `on_backpressure_buffer` on an `observable`. Here's an example that uses it: https://github.com/actor-framework/actor-framework/blob/main/examples/web_socket/stock-ticker.cpp#L122
4. Did you already have a look at the articles we have available on the flow API under https://www.interance.io/learning/? The `publisher`, for example, is covered in the `Concurrency` section of part 1. I hope the article also gives you sufficient details on the way buffers connect two actors. Looking at the example you have quoted now, maybe we shouldn't include that. The buffer resources are a rather low level API, definitely not how you would connect two actors in practice. The buffer resources are more of a building block that allow higher level abstractions on top. So that example in particular might lead people down the wrong road without sufficient context.

**Reply by phil-skillwon** [reply]
@Neverlord 
Yesterday, I tried using the caf::async::publisher<T> and stream<T> in high-frequency message processing scenarios, and they indeed provide a great experience. However, in low-frequency message scenarios, I believe that groups still have an irreplaceable role. 

I completely agree with what you said earlier: "Groups were meant as a loosely coupled pub/sub mechanism." Yes, in CAF-0.19.0, it’s very easy to create or obtain a group handle via `self->home_system().groups().get_local("xxx")`. This is a great loosely coupled design. By using `spawn_in_groups`, you can specify the relationship between an actor and multiple groups when creating it.

Especially for low-frequency message scenarios, such as when each group instance corresponds to a different chat channel, where most channels have low-frequency messages. There’s another scenario in industrial control, where sometimes you need to temporarily create a group instance to send special commands to multiple devices that share the same characteristics. In both of these cases, the message frequency is very low, and using a group is really suitable. On the other hand, when using `caf::async::publisher<T>`, you face a problem: passing the publisher handle from the outside to an actor is a composite action. In large projects, the logic of two `.cpp` files using the publisher might be quite distant, so it becomes troublesome to figure out how to store the publisher handle and pass it elegantly to multiple subscribers. Perhaps, in this case, decoupling using an API like `self->home_system().groups().get_local("xxx")` would be a better approach.

So, does this mean that groups will no longer be seen in future versions of CAF? 
Although the design of groups has flaws, its loosely coupled design is still great. Would it be possible to apply this loosely coupled design to `caf::async::publisher` so that we can use the publisher like a group without needing extra code? Just like what CAF-0.19.0 does with `self->home_system().groups().get_local("xxx")`.

**Reply by Neverlord** [reply]
> So, does this mean that groups will no longer be seen in future versions of CAF?

I'd definitely like to see an API for loose coupling in CAF at some point again. I don't expect that we can tackle that in the near future, but we'll keep this topic on our radar.

**Reply by phil-skillwon** [reply]
@Neverlord 
Thank you for your help. The transition from CAF 0.19.0 to CAF 1.0.2—such as changing from `send` to `mail`—isn’t really an issue. The main difficulty I’m facing is adjusting to the removal of the commonly used `group` data structure in CAF 1.0.2. However, based on your previous suggestions, I believe I’ve resolved the absence of `group`. I’m now using `caf::async::publisher<T>` and `caf::flow::multicaster<T>` to implement my own `MyGroup<T0, T1, T2, ...>` template class, where T0, T1, T2, etc. represent the message types supported by MyGroup. I do have one more question though: how can I implement backpressure using `caf::async::publisher<T>` and `caf::flow::multicaster<T>`, similar to how it’s done in the stream examples?

**Reply by Neverlord** [reply]
Just to be sure: you're not using the `multicaster` from multiple actors/threads? If you have an actor that receives messages and pushes them into the multicaster: that's fine (and the intended use).

What do you mean by "implement backpressure"? Flows always have backpressure in CAF. The multicaster also has member functions that return the current number of buffered items. You can use that information to "reject" inputs if you reach some upper limit (maybe that's what you meant?).

**Reply by phil-skillwon** [reply]
@Neverlord 
I think I understand what you mean now. The `multicaster<T>` exists as a property inside a `stateful_actor<T>`, and other actors obtain a `publisher<T>` through the `stateful_actor<T>`’s message handlers. In other words, multiple actors/threads only directly share the `publisher<T>`.

Below is a simple implementation of `MyGroup<T>`:
```cpp
CAF_BEGIN_TYPE_ID_BLOCK(MyTest, caf::first_custom_type_id)
CAF_ADD_ATOM(MyTest, GetMyGroupPublisherAtom)
CAF_ADD_ATOM(MyTest, JoinGroupAtom)
CAF_ADD_ATOM(MyTest, LeaveGroupAtom)
CAF_END_TYPE_ID_BLOCK(MyTest)

static constexpr size_t MAX_SIZE_MULTICASTER_BUF = 32;

template<typename T>
struct MyGroup 
{
    static_assert(std::is_move_constructible<T>::value, "T must be move constructible in class MyGroup");
    
    template<typename MsgType = T>
    explicit MyGroup(caf::event_based_actor* self, const std::string& name) 
        : m_self(self), m_name(name), m_multiCaster(self) 
    {
        m_publisher = m_multiCaster.as_observable().to_publisher();
    }
    MyGroup() = delete;
    ~MyGroup() = default;

    template<typename MsgType = T>
    caf::behavior make_behavior() 
    {
        return caf::behavior {
            [this](MsgType& msg) 
            {                
                if (!m_multiCaster.has_observers()) 
                {
                    return;
                }

                if (m_multiCaster.buffered() > MAX_SIZE_MULTICASTER_BUF) 
                {
                    return;
                }

                m_multiCaster.push(std::move(msg));
            },
            [this](GetMyGroupPublisherAtom&) -> caf::async::publisher<T> 
            {
                return m_publisher;
            },
            [this](const caf::exit_msg& msg) 
            {
                m_multiCaster.close();
                m_self->quit(msg.reason);
            },
            [this](const caf::error& msg) 
            {
                // ...
            },
            [this](caf::message& msg) 
            {
                // ...                
            },
        };
    }

    caf::event_based_actor* m_self = nullptr;
    std::string m_name;
    caf::flow::multicaster<T> m_multiCaster;
    caf::async::publisher<T> m_publisher;
};
```

In the simple implementation above, how does `multicaster<T>` achieve backpressure in high-throughput scenarios? Currently, I only check the number of buffered messages via `multicaster<T>::buffered()` and discard new messages if the threshold is reached, but this approach doesn’t feel quite right.

In the stream examples, I learned that backpressure can be configured by setting parameters such as:
```cpp
static constexpr auto MAX_BATCH_DELAY = 50ms;
static constexpr auto MAX_BATCH_SIZE = 10u;
static constexpr auto MAX_BUFFERED = 50u;
static constexpr auto DEMAND_THRESHOLD = 5u;

// Create the stream
auto stream = self->make_observable().from_callable(MyFunc).to_stream("xxx", MAX_BATCH_DELAY, MAX_BATCH_SIZE);

// Subscribe to the stream
auto disposableHandle = self->observe_as<caf::cow_vector<std::byte>>(stream, MAX_BUFFERED, DEMAND_THRESHOLD)
    .do_on_next([self](const caf::cow_vector<std::byte>& item) {
        // ...
    }).for_each([self](const caf::cow_vector<std::byte>& data) {
        // ...
    });
```
So, my question is: **<u>Can `caf::flow::multicaster<T>` and `caf::async::publisher<T>` implement backpressure in a way similar to streams</u>**.

Additionally, when reading the [Merging Flows Dynamically](https://www.interance.io/learning/cpp/guide/flows-part-2 "https://www.interance.io/learning/cpp/guide/flows-part-2")  section, I have a question. The code in that section is as follows:
```cpp
sys.spawn([](caf::event_based_actor* self) {
  self->make_observable()
    // Create the indexes for our sensors: 1, 2, 3.
    .iota(1)
    .take(3)
    // Turn the indexes into sensor flows and add the index to each value.
    .map([self](int index) {
      return open_sensor(self)
        .map([index](int value) { return std::pair{index, value}; })
        .as_observable();
    })
    // Turn `observable<observable<T>>` into `observable<T>` via merging.
    .merge()
    // Same as before: print the first ten readings to the console.
    .take(10)
    .for_each([self](std::pair<int, int> reading) {
      auto [index, value] = reading;
      self->println("sensor {}: {} °C", index, value);
    });
});
```

**<u>This code assumes that we know in advance how many flows need to be merged, in this case, the number of flows is three. Now, let's consider a scenario where we do not know the number of flows to be merged beforehand, and the merging operation is not a one-time action. For example, we might merge three flows now, but a few seconds later we might need to merge five flows, and a few minutes later we might need to merge eight flows. How would we implement this scenario</u>**?

Lastly, about `multicaster<T>` and streams: **<u>`multicaster<T>` replicates the same data and distributes it to multiple subscribers, while a stream distributes a given data item (or batch) to only one of the subscribers. In other words, `multicaster<T>` copies data, but streams do not, am I understanding this correctly</u>**?

**Comment by Neverlord** [comment]
> Can caf::flow::multicaster<T> (...) implement backpressure in a way similar to streams.

It does participate in backpressure. However, you are feeding in regular actor messages. Actor messages do not (and cannot) implement backpressure. You are already checking whether the buffered amount of messages exceeds a certain limit, which is the best thing to do in this situation. You could at most return an `error` to the sender of a message in case the buffer is full. That would force senders to terminate by default (unless the provide a custom error handler).

> we might merge three flows now, but a few seconds later we might need to merge five flows, and a few minutes later we might need to merge eight flows. How would we implement this scenario?

If you are feeding in flows that will be produced later, you can once again use a `multicaster`. Only this time, `multicaster<observable<T>>` and then you can push new flows.

> ... multicaster and streams ...

A `stream` is similar to a `publisher`. The `multicaster` has nothing to do with that. A `multicaster` just represents an observable and thus can be turned into either `stream` or `publisher`, depending on your use case.

**Reply by phil-skillwon** [reply]
@Neverlord 
I'm sorry for the confusion in my previous question. The question I mentioned:

```bash
we might merge three flows now, but a few seconds later we might need to merge five flows, and a few minutes later we might need to merge eight flows. How would we implement this scenario?
```

is actually a separate question and unrelated to `multicaster<T>`/`publisher<T>`.

When reading [Merging Flows Dynamically](https://www.interance.io/learning/cpp/guide/flows-part-2), I came across the following code snippet:

```cpp
sys.spawn([](caf::event_based_actor* self) {
  self->make_observable()
    // Create the indexes for our sensors: 1, 2, 3.
    .iota(1)
    .take(3)
    // Turn the indexes into sensor flows and add the index to each value.
    .map([self](int index) {
      return open_sensor(self)
        .map([index](int value) { return std::pair{index, value}; })
        .as_observable();
    })
    // Turn `observable<observable<T>>` into `observable<T>` via merging.
    .merge()
    // Same as before: print the first ten readings to the console.
    .take(10)
    .for_each([self](std::pair<int, int> reading) {
      auto [index, value] = reading;
      self->println("sensor {}: {} °C", index, value);
    });
});
```

My question is: suppose the usage scenario is different from the code above. Initially, we merge a fixed set of 3 flows, but a few seconds later, we have 5 new flows that need to be merged, and a few minutes after that, we have 8 more flows to merge. If the situation changes dynamically over time like this, how can we use flows to handle such a scenario?

**Reply by Neverlord** [reply]
> how can we use flows to handle such a scenario?

One way is as I said:

> If you are feeding in flows that will be produced later, you can once again use a multicaster. Only this time, `multicaster<observable<T>>` and then you can push new flows.

This allows you to call `push(...)` at some later time with a fresh observable to be merged with the others. Doesn't have to be a `multicaster` of course. In general, you need to produce a flow ob observables. The `multicaster` is one way to set this up if flows are produced later on through some external event (such as incoming actor messages). Sounds like that fits your use case.

**Reply by phil-skillwon** [reply]
@Neverlord 
If the flow computation has already started, but new observables are generated later on (for example, three new observables after X seconds, and eight new observables after Y minutes), how can I merge these newly created observables into the ongoing flow computation?

In other words, the flow computation has already begun, as shown in the code below:

```C++
self->observe_as<caf::cow_vector<std::byte>>(stream, MAX_BUFFERED, DEMAND_THRESHOLD)
    .do_on_next([this](const caf::cow_vector<std::byte>& item) {
        // ...
    })
    .for_each([this, stream](const caf::cow_vector<std::byte>& data) {
        // ...
    });
```

After X seconds, I have three newly created stream instances. How can I merge them into the already ongoing flow computation above?

After Y minutes, I get eight more newly created stream instances. How can I also merge those into the already ongoing flow computation above?

**Reply by Neverlord** [reply]
I assume you receive those streams via message?

Given a member variable `inputs` of type `multicaster<observable<cow_vector<std::byte>>`, we declare our pipeline once during setup:


```
  inputs.as_observable().merge().for_each(...);
```

When receiving a new stream, we can merge them into the already ongoing flow via:

```
  inputs.push(self->observe_as<caf::cow_vector<std::byte>>(stream, MAX_BUFFERED, DEMAND_THRESHOLD))
```

**Reply by phil-skillwon** [reply]
@Neverlord 
I fully understand what you mean now. Yesterday, I implemented it using `multicaster<T>`. I always thought that the `merge` template function had to take arguments. It turns out I missed some important details while reading the "Merging Flows Dynamically" section before. Thanks for your help.
